### PR TITLE
(NPUP-6) Implement function evaluation.

### PR DESCRIPTION
**Pull request by peterhuene**
These commits implement the following:

* Puppet function evaluation.  Function expressions are now registered with the dispatcher.
* The `--trace` option for the compile command: enables printing out Puppet backtraces.  Native frames (Puppet functions implemented in native code) show up on the backtrace without source information.
* Private module functions (experimental).  Functions in modules can be marked with the 'private' keyword and the caller must come from the same module or else an error is raised.

Most of the churn comes from amending all of the places where `evaluation_exception` is thrown to capture and store the current Puppet backtrace.

Additionally, the `call_evaluator` was removed in favor of discrete evaluators for functions, classes, defined types, and nodes as the former was pretty confusing ("calling" a defined type, for instance).

### CODE REVIEW

**Comment by peterhuene** [comment]
@hlindberg This uber-low-priority PR is ready whenever you have spare cycles / are bored.

**Comment by hlindberg** [comment]
Note that "calling a type" is going to be possible soon in the Ruby implementation. The meaning of that is to create a new instance of the type. e.g.  Array({a => 10, b => 20}) creates an array from the hash.

File: lib/src/compiler/scanner.cc [hunk 1]
```diff
@@ -439,6 +439,17 @@ namespace puppet { namespace compiler {
             );
         }
 
+        // Ensure private functions have an associated module
+        if (expression.is_private && expression.tree && !expression.tree->module()) {
+            throw parse_exception(
+                (boost::format("function '%1%' cannot be declared private because it is not contained in a module.") %
```
Comment (hlindberg) [inline lib/src/compiler/scanner.cc]: You can have functions in the environment too, if they are private they cannot be called from a module (only from puppet itself, and from other functions, and site.pp). (Albeit we have not specified this yet, but that is what is expected in this case). Suggest logging a separate ticket, as the support for private is a exploratory at this point.

File: lib/tests/fixtures/compiler/evaluation/bare_word.pp [hunk 2]
```diff
@@ -0,0 +1,36 @@
+notice a-b
+notice _foo-bar_-baz
+notice _
+
+$a = a-b
```
Comment (hlindberg) [inline lib/tests/fixtures/compiler/evaluation/bare_word.pp]: Is there a test for name ending with hyphen `-` (not allowed)

**Comment by hlindberg** [comment]
(There was a late change, and I am not sure if that was for addressing my comment or not.)
Basically, the runtime should treat the environment as if it was a module named "environment" in addition to representing the global namespace "::". Thus, there can be namespaced functoins in an environment named something like `environment::foo` (they are not named after the actual environment because they change names between test, production, ticket1234, etc). As an example, the `environment::data()` functions can be used to define lookup data. It is not a hard requirement that an environment namespaces its functions with "environment", it may contain functions in the global namespace as well. 

Suggest a separate ticket to deal with environment related functions and their names.

File: lib/tests/fixtures/compiler/evaluation/bare_word.pp (hunk 2 above)
Comment (peterhuene) [inline lib/tests/fixtures/compiler/evaluation/bare_word.pp]: I will add one.

**Comment by peterhuene** [comment]
I'll file a ticket to handle private environment functions and the `environment` namespace.

File: lib/src/compiler/scanner.cc (hunk 1 above)
Comment (peterhuene) [inline lib/src/compiler/scanner.cc]: Weird, I didn't push anything up today :/  I have no idea how this became an outdated comment.

**Pull request merged**